lib/internal/filter.js

```diff
@@ -1,31 +1,19 @@
-import arrayMap from 'lodash/_arrayMap';
-import property from 'lodash/_baseProperty';
+import filter from 'lodash/filter';
 import noop from 'lodash/noop';
 import once from './once';
 
 export default function _filter(eachfn, arr, iteratee, callback) {
     callback = once(callback || noop);
-    var results = [];
+    var truthValues = new Array(arr.length);
     eachfn(arr, function (x, index, callback) {
         iteratee(x, function (err, v) {
-            if (err) {
-                callback(err);
-            }
-            else {
-                if (v) {
-                    results.push({index: index, value: x});
-                }
-                callback();
-            }
+            truthValues[index] = !!v;
```

Comment:
so for objects we're assigning `x[key]` where x is an array. Thus creating a sparse array. I think I'd be more comfortable making `x` an object.